arrow/src/compute/kernels/arithmetic.rs

```diff
@@ -774,6 +778,110 @@ pub fn add_dyn(left: &dyn Array, right: &dyn Array) -> Result<ArrayRef> {
         DataType::Dictionary(_, _) => {
             typed_dict_math_op!(left, right, |a, b| a + b, math_op_dict)
         }
+        DataType::Date32 => {
+            let l = left
+                .as_any()
+                .downcast_ref::<PrimitiveArray<Date32Type>>()
+                .ok_or_else(|| {
+                    ArrowError::CastError(format!(
+                        "Left array cannot be cast to Date32Type",
+                    ))
+                })?;
```

Comment:
For what it is worth, most of the rest of the codebase simply `panic`s if the types don't match (as it is a fairly large runtime invariant).

So a common pattern is like:

```suggestion
            let l = left
                .as_any()
                .downcast_ref::<PrimitiveArray<Date32Type>>()
                .unwrap();
```

There is also https://docs.rs/arrow/17.0.0/arrow/array/fn.as_primitive_array.html# to make things even simpler: 

```
            let l = as_primitive_array<Date32Type>(&left);
```

Reply:
Thank you, the verbosity was bugging the hell out of me.